### PR TITLE
oidc oauth reverse proxy sidecar

### PR DESCRIPTION
**Pull request by shihchang**
Add oidc oauth reverse proxy sidecar to elasticsearch, prometheus, kibana, and grafana
* redirect unauthenticated browser access to keycloak to login
* require Bearer type authorization header for API access
* continue to support Basic auth for API access

### CODE REVIEW

**Comment by shihchang** [comment]
The oidc proxy implementation is the lua scripts in [pkg/constants/oidc_proxy.go](https://github.com/verrazzano/verrazzano-monitoring-operator/pull/63/files#diff-f5b45aa5191c24022e072f8901214bae54a32904c90bc6b9a2b42a01abb09f45), embedded in a configMap to configure a nginx sidecar attached to each of the deployments.

File: pkg/constants/oidc_proxy.go [hunk 1]
```diff
@@ -0,0 +1,540 @@
+// Copyright (C) 2020, 2021, Oracle and/or its affiliates.
```
Comment (pfmackin) [inline pkg/constants/oidc_proxy.go:1]: remove 2020

File: pkg/constants/oidc_proxy.go [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+// Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+package constants
+
+//Define the configuration templates for OIDC proxy
+const (
+	OidcCallbackPath = "/_authentication_callback"
```
Comment (pfmackin) [inline pkg/constants/oidc_proxy.go:8]: formatting doesn't line up

File: pkg/constants/oidc_proxy.go [hunk 3]
```diff
@@ -0,0 +1,540 @@
+// Copyright (C) 2020, 2021, Oracle and/or its affiliates.
+// Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+package constants
+
+//Define the configuration templates for OIDC proxy
```
Comment (pfmackin) [inline pkg/constants/oidc_proxy.go:6]: add space after //

File: pkg/constants/oidc_proxy.go (hunk 2 above)
Comment (pfmackin) [inline pkg/constants/oidc_proxy.go:8]: Where is this template coming from?   Can you put a doc link in the code?

File: pkg/resources/deployments/deployment.go [hunk 4]
```diff
@@ -119,6 +122,11 @@ func New(vmo *vmcontrollerv1.VerrazzanoMonitoringInstance, operatorConfig *confi
 		deployment.Spec.Template.Spec.SecurityContext = &corev1.PodSecurityContext{
 			FSGroup: &grafanaGid,
 		}
+		if config.Grafana.OidcProxy != nil {
+			oidcConfig, oidcProxy := resources.CreateOidcProxy(vmo, &vmo.Spec.Grafana.Resources, &config.Grafana)
```
Comment (pfmackin) [inline pkg/resources/deployments/deployment.go:126]: oidcConfig should be called oidcVolume to make it clear what it is.  Same comment throughout.

File: pkg/resources/ingresses/ingress.go [hunk 5]
```diff
@@ -198,3 +214,55 @@ func noAuthOnHealthCheckSnippet(vmo *vmcontrollerv1.VerrazzanoMonitoringInstance
 }
 `
 }
+
+//oidcProxyIngress creates the Ingress of the OidcProxy
+func oidcProxyIngress(vmo *vmcontrollerv1.VerrazzanoMonitoringInstance, component *config.ComponentDetails) *extensions_v1beta1.Ingress {
```
Comment (pfmackin) [inline pkg/resources/ingresses/ingress.go:219]: This should be newOidcProxyIngress or createOidcProxyIngress

File: pkg/resources/ingresses/ingress.go [hunk 6]
```diff
@@ -119,26 +119,20 @@ func New(vmo *vmcontrollerv1.VerrazzanoMonitoringInstance) ([]*extensions_v1beta
 	ingresses = append(ingresses, ingress)
 
 	if vmo.Spec.Grafana.Enabled {
-		// Create Ingress Rule for Grafana Endpoint
-		ingRule = createIngressRuleElement(vmo, config.Grafana)
-		host := config.Grafana.Name + "." + vmo.Spec.URI
-		ingress, err := createIngressElementNoBasicAuth(vmo, host, config.Grafana, ingRule)
-		if err != nil {
-			return ingresses, err
+		if config.Grafana.OidcProxy != nil {
```
Comment (michael-cico) [inline pkg/resources/ingresses/ingress.go:122]: It mostly seems pretty straightforward, but what sets the OidcProxy configuration in the configmap for each component?  Is that going to be done by the `verrazzano-operator`?

Comment (shihchang) [reply pkg/resources/ingresses/ingress.go:122]: In pkg/config/components.go controlled by the OIDC_AUTH_ENABLED flage of env-var

File: pkg/constants/oidc_proxy.go (hunk 1 above)
Comment (shihchang) [reply pkg/constants/oidc_proxy.go]: removed

File: pkg/constants/oidc_proxy.go (hunk 2 above)
Comment (shihchang) [reply pkg/constants/oidc_proxy.go]: fixed

File: pkg/constants/oidc_proxy.go (hunk 3 above)
Comment (shihchang) [reply pkg/constants/oidc_proxy.go]: fixed

File: pkg/constants/oidc_proxy.go (hunk 2 above)
Comment (shihchang) [reply pkg/constants/oidc_proxy.go]: added godoc comment to each const items

File: pkg/resources/deployments/prometheus.go [hunk 7]
```diff
@@ -186,6 +188,12 @@ func createPrometheusNodeDeploymentElements(vmo *vmcontrollerv1.VerrazzanoMonito
 		}
 		prometheusDeployment.Spec.Template.Spec.Volumes = append(prometheusDeployment.Spec.Template.Spec.Volumes, volume)
 
+		if config.Prometheus.OidcProxy != nil {
+			oidcConfig, oidcProxy := resources.CreateOidcProxy(vmo, &vmo.Spec.Prometheus.Resources, &config.Prometheus)
```
[the remark this answers is not in the document]
Comment (shihchang) [reply pkg/resources/deployments/prometheus.go]: fixed

File: pkg/resources/ingresses/ingress.go (hunk 5 above)
Comment (shihchang) [reply pkg/resources/ingresses/ingress.go]: chagned to newOidcProxyIngress

**Review by michael-cico — APPROVED** [review]

**Review by pfmackin — APPROVED** [review]

**Review by wmhopkins — APPROVED** [review]
I'm approving this, but I have concerns about the approach to implementing the proxy.
- I'm not convinced that we've implented the OIDC protocols correctly/completely -- not saying we haven't, just that it needs a careful review to make sure, and I haven't had time to do that.
- I don't think we should be storing JWTs, or any part of them, in a cookie. If we really need to support sessions, we need to implement a session cookie. It may also be possible (and much better) to let Keycloak manage the sessions so we don't have to.
- With both this proxy and the api-proxy, I'm concerned that the code/configuration for nginx is potentially exposed because it's in a config map (including user accounts and passwords). Would be safer to build our own proxy image, I think, rather than rely on the generic one configured dynamically. We should also consolidate on a single proxy impl for both api-proxy and oidc-proxy.

Can you create a jira to look at these, or add them to the existing jira for authZ and rename it (to include more than authZ)?

**Pull request merged**